### ISSUE TITLE
Craft special cases : handle custom recipes and not consumed items

### ISSUE THREAD

**Issue by rom1504** [andrewrk/mineflayer#222]
Currently it is assumed that all the blocks used in a recipe are consumed (https://github.com/andrewrk/mineflayer/blob/master/lib/plugins/inventory.js#L572).

It is not true for all recipes (see http://minecraft.gamepedia.com/Written_Book )

So it should be handled by mineflayer : 
* either by adding a consumed/not consumed in the recipes enum
* either in figuring out whether the server send us the new crafting grid items (or absence of item) 

Custom recipes are related to this consumption/not consumption

**Comment by roblabla** [comment]
We should test the proxy with a bukkit plugin like http://dev.bukkit.org/bukkit-plugins/recipemanager/ to figure out how it's done.

**Comment by deathcap** [comment]
Also consider "container" items, e.g., milk bucket's container = empty bucket, returned from crafting cake http://minecraft.gamepedia.com/Cake

**Comment by rom1504** [comment]
Yeah so.. I finally understoof what is "outShape" in recipes.json
inShape is the shape before crafting, outShape is the shape after crafting

So, it should be checked if there is an outShape there https://github.com/andrewrk/mineflayer/blob/master/lib/plugins/inventory.js#L572 and I think that will handle all the cases we are talking about in that issue.

**Comment by rom1504** [comment]
So, it would be nice to test if we can craft a cake. It that works, this issue is half-done.

**Comment by rom1504** [comment]
Ok crafting a cake works.

**Comment by rom1504** [comment]
Is it possible to get a "written book" by the creative inventory ? it doesn't seem there so I can't test whether that recipe would work.

**Comment by thejoshwolfe** [comment]
creative api can probably give you everything, since it's how the normal client simply moves items around.

**Comment by deathcap** [comment]
Also what about firework crafting recipes (according to http://minecraft.gamepedia.com/Fireworks "firework rocket is one of the few items in the game that cannot be taken directly out of a creative inventory") and [item repair recipes](http://minecraft.gamepedia.com/Repair) (combining two damaged tools)

**Comment by rom1504** [comment]
I meant "creative inventory" not creative API, ie I don't know how to give myself a written book in the actual game.
I'll try the firework recipe.

**Comment by rom1504** [comment]
crafting fireworks works.

**Comment by rom1504** [comment]
I'd better have an automatic test for this though (see https://github.com/andrewrk/mineflayer/issues/239#issuecomment-87342357) instead of testing manually.

**Comment by deathcap** [comment]
> I meant "creative inventory" not creative API, ie I don't know how to give myself a written book in the actual game.

You can use commands:

``
/give player minecraft:written_book 1 0 {author:'Player', title:'title1', resolved:1, generation:0, pages: ['hi']}
```

**Comment by rom1504** [comment]
Ok that worked.

That recipe is broken in mineflayer, the reason is this recipe has ingredients and "outIngredients", which is not handled by mineflayer (and minecraft-data).

see https://github.com/PrismarineJS/minecraft-data/issues/10

**Comment by rom1504** [comment]
Recipes with metadata (even the simplest one like "white" log --> "white" wood) are broken. The problem is likely in the data, fixing this depends on https://github.com/PrismarineJS/minecraft-data/issues/7

Also : it would be nice to know which recipe is broken and which is not (see https://github.com/andrewrk/mineflayer/issues/239)

**Comment by Matitron** [comment]
with 1.17 (note:it may have happened before 1.17) but containers like bottles that are used eg: if i make a recipie that turns a honey bottle into a potato the bottle will remain but we still can't control witch item is consumed but its a start i suppose :P
